common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+#ifndef socket_refactoring_H
+#define socket_refactoring_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <linux/if.h>
+#include <linux/if_packet.h>
+#include <linux/netlink.h>
+#include <linux/un.h>
+
+#include <algorithm>
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+#include <tlvf/CmduMessageRx.h>
+#include <tlvf/common/sMacAddr.h>
+#include <tlvf/ieee_1905_1/eMessageType.h>
+
+#include <mapf/common/utils.h>
+
+#include "beerocks_event_loop.h"
+
+namespace beerocks {
+namespace sockets {
+
+/**
+ * Array of bytes used to hold data received through a socket.
+ * Code is programmed to interfaces so it does not care about which implementation is used.
+ * Unit tests can use a mock and set different expectations per test (pretend that different data
+ * has been received through the socket).
+ */
+class Buffer {
+public:
+    virtual ~Buffer()                   = default;
+    virtual const uint8_t *data() const = 0;
+    virtual size_t size() const         = 0;
+
+    uint8_t *data() { return const_cast<uint8_t *>(const_cast<const Buffer *>(this)->data()); }
+};
+
+/**
+ * First example implementation
+ */
+class BufferImpl1 : public Buffer {
+public:
+    BufferImpl1(uint8_t *data, size_t size) : m_data(data), m_size(size) {}
+    const uint8_t *data() const override { return m_data; }
+    size_t size() const override { return m_size; }
+
+private:
+    uint8_t *m_data;
+    size_t m_size;
+};
+
+/**
+ * Second example implementation
+ */
+template <size_t Size> class BufferImpl2 : public Buffer {
+public:
+    const uint8_t *data() const override { return m_data; }
+    size_t size() const override { return sizeof(m_data); }
+
+private:
+    uint8_t m_data[Size];
+};
+
+/**
+ * This interface models OS resources implementing the file descriptor interface.
+ */
+class FileDescriptor {

```

Comment:
AFAIK we use the following calls to open a file descriptor:
- socket()
- socketpair()
- timerfd_create()
- epoll_create()

I may even have missed some...

One option would be to create static members for each of those in the FileDescriptorImpl class. I'm kind of a fan of that approach. It does mean we can't treat the different kinds of file descriptors as different types, which we probably do want.

Note that timerfd_create and epoll_create are only used in EventLoop which anyway deals with raw file descriptors. So indeed, we could simplify the code by having *only* a Socket class and no distinct FileDescriptor class.

Reply:
My next movement is to create a `Timer` class to deal with periodic tasks and timeouts. This class will include a file descriptor obtained by calling `timerfd_create()`. This file descriptor will be used to install handlers in EventLoop, like we do with sockets. 
We could use an `int` in calls to `EventLoop::registerHandler()`, but you convinced me to create a class including `close()` on destructor and deleted copy constructor and assignment operator (to avoid errors by design). 
I think creating static members in FileDescriptorImpl class violates the [Open-closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle) and we should avoid it.

This subject is very important and maybe it deserves a special meeting to discuss. Due to your comments, I've got the feeling that I haven't been able to transmit all the necessities and implications to you.